README.md
# gatsby-plugin-segment-js

A lightweight & feature-rich Gatsby plugin to easily add [Segment JS snippet](https://segment.com/docs/connections/sources/catalog/libraries/website/javascript/quickstart/) to your site.

## We need your help! 🙏🏽
We're looking for active contributors to this repo!  If you're interested, simply open an issue or PR of your own and indicate that you'd like to help.  Check out our open issues and PRs.  We also need to beef up testing.  Contributors get to:
- manage versioning and deploys by publishing new versions to NPM
- determine which features get launched by merging pull requests
- oversee the community through commenting on and managing pull requests and issues

⚜️ Being an active contributor is great for the community and your engineering resume.⚜️

## Features

Packed with features:

- use multiple write keys (one for prod env, another optional one for dev)
- disable page view tracking (just in case you want to add it later manually)
- will use a default Segment code snippet (currently `v4.15.3`), or you can provide your own custom snippet

## Install

- NPM: `$ npm install --save gatsby-plugin-segment-js`
- YARN: `$ yarn add gatsby-plugin-segment-js`

## How to use

### Setup

In your gatsby-config.js file:

```javascript
plugins: [
  {
    resolve: `gatsby-plugin-segment-js`,
    options: {
      // your segment write key for your production environment
      // when process.env.NODE_ENV === 'production'
      // required; non-empty string
      prodKey: 'SEGMENT_PRODUCTION_WRITE_KEY',

      // if you have a development env for your segment account, paste that key here
      // when process.env.NODE_ENV === 'development'
      // optional; non-empty string
      devKey: 'SEGMENT_DEV_WRITE_KEY',

      // Boolean indicating if you want this plugin to perform any automated analytics.page() calls
      // at all, or not.
      // If set to false, see below on how to track pageviews manually.
      // 
      // This plugin will attempt to intelligently prevent duplicate page() calls.
      // 
      // Default: true
      trackPage: true,

      // Boolean indicating if you want this plugin to perform a page() call immediately once the snippet
      // is loaded.
      // 
      // You might want to disable this if you *only* want page() calls to occur upon Client-side routing
      // updates. See `trackPageOnRouteUpdate` option.
      // 
      // This plugin will still attempt to intelligently prevent duplicate page() calls.
      // 
      // Default: true
      trackPageImmediately: true,

      // Boolean indicating whether to ignore `page` calls by this plugin before we call `analytics.load`
      // and/or the `ready` event has been emitted by `analytics`.
      // 
      // Useful in some cases to prevent issues "queuing" `page` events before Segment is fully loaded.
      trackPageOnlyIfReady: false,

      // Boolean indicating whether to perform any page() calls during Client-side routing updates.
      // 
      // You might want to disable `trackPageImmediately` if you *only* want page() calls to occur upon
      // Client-side routing updates.
      // 
      // This plugin will still attempt to intelligently prevent duplicate page() calls.
      // 
      // Default: true
      trackPageOnRouteUpdate: true,

      // Number indicating what delay (in ms) should be used when calling analytics.page() in response
      // to a Gatsby `onRouteUpdate` event. This can help to ensure that the segment route tracking is 
      // in sync with the actual Gatsby route. Otherwise you can end up in a state where the Segment
      // page tracking reports the previous page on route change.
      // 
      // See https://www.gatsbyjs.com/docs/reference/config-files/gatsby-browser/ for more information.
      // 
      // Default: 50
      trackPageOnRouteUpdateDelay: 50,

      // Boolean indicating whether or not to add the document.title as the first argument to
      // the analytics.page() calls. Segment uses some sane defaults, but some users of this plugin
      // have wanted to do this in the past.
      // 
      // E.g `analytics.page(document.title)`
      // 
      // Default: false
      trackPageWithTitle: false,
      
      // Boolean indicating whether to call analytics.load() immediately, or to delay it by a specified
      // number of ms. Can be useful if you want to wait a specifc amount of time before calling
      // analytics.load() and kicking off the activity that occurs with that call.
      // 
      // Default: false
      delayLoad: false,

      // Number indicating (in ms) how long to wait for before analytics.load() will be called if
      // the `delayLoad` option is set to `true`.
      // 
      // Default: 1000
      delayLoadDelay: 1000,

      // Boolean indicating whether to delay calling analytics.load() until either:
      // 1) The User interacts with the page by scrolling
      // OR
      // 2) The User triggers a Gatsby route change.
      // 
      // The `delayLoad` option can be used in addition to this option.
      // 
      // NOTE: 
      // The route change will only be triggered if you leverage client-side routing (ie, Gatsby <Link>)
      // So if you leverage server-side routing with this feature, only a User scroll will trigger
      // the `load` call. This is because client-side routing does not do
      // a full page refresh, but server-side routing does. Therfore server-side routing will never
      // appear to have been triggered by a User interaction.
      // 
      // This is an advanced feature most often used to try to help improve your website's TTI (for SEO, UX, etc).
      // 
      // See here for more context:
      // Client-side routing and browser code: https://www.gatsbyjs.com/docs/reference/config-files/gatsby-browser/
      // GIF: https://github.com/benjaminhoffman/gatsby-plugin-segment-js/pull/19#issuecomment-559569483
      // TTI: https://github.com/GoogleChrome/lighthouse/blob/master/docs/scoring.md#performance
      // Problem/solution: https://marketingexamples.com/seo/performance
      // 
      // Default: false
      delayLoadUntilActivity: false,

      // Number indicating (in ms) any additional delay to wait before calling analytics.load()
      // after User "activity" has occurred in conjunction with the `delayLoadUntilActivity` option.
      // 
      // Default: 0
      delayLoadUntilActivityAdditionalDelay: 0,

      // Whether to completely skip calling `analytics.load({writeKey})`.
      // ADVANCED FEATURE: only use if you are calling `analytics.load({writeKey})` manually
      // elsewhere in your code or are using a library
      // like: https://github.com/segmentio/consent-manager that will call it for you.
      // Useful for only loading the tracking script once a user has opted in to being tracked, for example.
      // 
      // Default: false
      manualLoad: false,

      // If you need to proxy events through a custom endpoint,
      // add a `host` property (defaults to https://cdn.segment.io)
      // Segment docs:
      //   - https://segment.com/docs/connections/sources/custom-domains
      //   - https://segment.com/docs/connections/sources/catalog/libraries/website/javascript/#proxy
      // 
      // Default: 'https://cdn.segment.io'
      host: 'https://override-segment-endpoint',

      // This package will use a default version of Segment's code snippet, but
      // if you'd like to include your own you can do so here. This is useful if
      // the version this package uses is different than the one you'd like to
      // use...or you need to do something custom.
      // While you should NOT use a back-ticked template string here, the string
      // will be evaluated as template literal with the following variables
      // available to it:
      //    - `writeKey`: The appropriate value from the `prodKey` and `devKey`
      //      options, based on the `NODE_ENV`
      //    - any of the other options passed here
      // 
      // NOTES: 
      // - If you provide a custom snippet, an immediate call to
      //   `analytics.load()` and/or `analytics.page()` will not be added by
      //   this plugin. You can - of course - add them yourself to your snippet.
      // - If your custom snippet does not include a call to `analytics.load()`
      //   then you must either:
      //   1. Manually load it and set the `manualLoad` option here to `true`
      //   2. Use the `delayLoad` option here
      customSnippet: '!function(){var analytics=window.analytics||[];...;analytics.load("${writeKey}");analytics.page();}}();'
    }
  }
];
```

### Loading

A typical Segment setup using this plugin will add an initial "snippet" for Segment to the page. Once that snippet is there, it needs to be "loaded" completely by making a call to `analytics.load(<your segment key here>)` before it will actually send over any events to Segment. This `load` call causes a flurry of XHR calls, and some people seem to be quite concerned with having those calls not occur until after some delay - usually for SEO scores, though it's not clear to me if the way Segment handles `load` calls will impact TTI, etc. Regardless, if you want this plugin to delay calling `load`, you have 2 options:
    1. `delayLoad`: This will cause a straightforward delay before `load` is called. Use in conjunction with `delayLoadDelay` to tweak the amount of the delay.
    2. `delayLoadUntilActivity`: This will wait to call `load` until __either__ (a) the user triggers a "scroll" event or (b) the user triggers route change. This option will take precendence over the `delayLoad` option.

### Track Events

If you want to track events, you simply invoke Segment as normal in your React components — (`window.analytics.track('Event Name', {...})` — and you should see the events within your Segment debugger! For example, if you wanted to track events on a click, it may look something like this:

```javascript
class IndexPage extends React.Component {
    ...
    _handleClick() {
        window.analytics.track("Track Event Fired", {
            userId: user.id,
            gender: 'male',
            age: 33,
        });
    }
    render() {
        return (
            <p>
                <Link onClick={this._handleClick} to="/">
                    Click here
                </Link>{" "}
                to see a track event
            </p>
        );
    }
}
```

### Track Pageviews

If you want to track pageviews automatically, set (or leave) `trackPage` to `true` in your `gatsby-config.js` file. What we mean by _"automatically"_ is that we will intelligently decide how or when to invoke `window.analytics.page()` for you.

This involves identifying if the last `page` call was made while the User was on the same `location.pathname` as the previous `page` call. If they appear to be the same, we will skip the duplicate call.

This plugin can also leverage Gatsby's `onRouteUpdate` API in the `gatsby-browser.js` file ([link][on-route-update]) if you so desire. This is accomplished by setting (or leaving) the `trackPageOnRouteUpdate` to `true`. If this option is enabled, we will make a `page` call every time that the `onRouteUpdate` handler is called. That includes the initial route, and any route changes.

If you only want to use this plugin to inject and load Segment, and not make any `page` calls, this can be done. Just set `trackPage: false`. But then you will have to make `page` calls yourself in your `gatsby-browser.js` file, like this:

```javascript
// gatsby-browser.js
exports.onRouteUpdate = () => {
  window.analytics && window.analytics.page();
};
```

[on-route-update]: https://www.gatsbyjs.org/docs/browser-apis/#onRouteUpdate

CHANGELOG.md
## 5.0.0
- BREAKING CHANGE: The `options` now support more complex things, and some thing are renamed. Please check out all of the options in the `README`.

## 4.0.3
- Ensure that cb is a function before calling it. https://github.com/benjaminhoffman/gatsby-plugin-segment-js/pull/57

## 4.0.2
- Fix syntax error in delayed loader code. https://github.com/benjaminhoffman/gatsby-plugin-segment-js/pull/56

## 4.0.1
- Better idempotency. https://github.com/benjaminhoffman/gatsby-plugin-segment-js/pull/54

## 4.0.0

- Add `includeTitleInPageCall` option, and set it to `true` by default. Strictly speaking, this was a "breaking change", hence the new major version.
- Fix bug where `analytics.page()` would be called twice.

## 3.7.2

- Updated `gatsby` peer dependency to support `^2 || ^3 || ^4 || ^5` https://github.com/benjaminhoffman/gatsby-plugin-segment-js/pull/52

## 3.7.1

- Updated `gatsby` peer dependency to support `^2 || ^3 || ^4` https://github.com/benjaminhoffman/gatsby-plugin-segment-js/pull/43

## 3.7.0

- Updated default snippet to `4.13.2`
- Added `customSnippet` option to allow for providing arbitrary snippet code
- Decreased NPM package size
- Some small optimizations
- Added linting for developers

- Added correct peer dependencies(thanks @LekoArts) and updated readme to show correct usage of analytics.load(thanks @seankovacs)

## 3.6.1

- Added correct peer dependencies(thanks @LekoArts) and updated readme to show correct usage of analytics.load(thanks @seankovacs)

## 3.6.0

- Added `trackPageDelay` option that allows specifying the amount of time to wait before calling `trackPage`. Thanks to @sreucherand!

## 3.5.0

- Added `addSourceMiddleware`, `addIntegrationMiddleware`, `setAnonymousId`, and `addDestinationMiddleware` methods to Segment snippet

## 3.4.0

- Adds `manualLoad` plugin option to allow users to manually load Segment snippet, e.g. waiting for a user to opt into being tracked for GDPR

## 3.3.1

- Send `document.title` with page tracking events for enhanced event data

## 3.3.0

- Adds a 50ms delay to page tracking calls to prevent erroneously sending duplicate page events

## 3.2.1

- Delete package-lock.json

## 3.2.0

- Bug: Built files not uploaded to npm

## 3.1.1

- Updated README.md

## 3.1.0

- NEW FEATURE: we've added one new feature with two options: `delayLoad` and its friend `delayLoadTime`. When the former is set to true, we will delay loading Segment `delayLoadTime` seconds _after_ either a route change or user page scroll, whichever comes first. This functionality was built to help with SEO, UX, etc by preventing Segment from increasing your TTI.

## 3.0.0

- BREAKING CHANGE: you must now explicitly pass `trackPage: true` in your `gatsby-config.js` file if you want us to automatically track pageviews

- BREAKING CHANGE: previously we would only fire `analytics.page()` on the initial page load. But now we will invoke it on each route transition. See README for more details.

- Added example dir

- Expanded README.md and adds CHANGELOG.md

## 2.1.2

- Bumps Segment snippet from 4.0.0 to 4.1.0

## 2.0.0

- Basic version working


src/gatsby-ssr.js
// https://www.gatsbyjs.com/docs/reference/config-files/gatsby-ssr/
import React from "react";

export function onRenderBody({ setHeadComponents }, pluginOptions) {
  const {
    prodKey,
    devKey,
    host = "https://cdn.segment.io",
    trackPage = true,
    trackPageImmediately = true,
    trackPageOnlyIfReady = false,
    // This is ONLY for the Browser side code
    // trackPageOnRouteUpdate = true,
    // trackPageOnRouteUpdateDelay = 50,
    includeTitleInTrackPage,
    manualLoad = false,
    delayLoad = false,
    delayLoadDelay = 1000,
    delayLoadUntilActivity = false,
    delayLoadUntilActivityAdditionalDelay = 0,
    customSnippet,
  } = pluginOptions;

  let {
    trackPageWithTitle = false,
  } = pluginOptions;

  if (typeof includeTitleInTrackPage === 'boolean') {
    console.warn('WARNING: option for gatsby-plugin-segment "includeTitleInTrackPage" is deprecated. Please use "trackPageWithTitle" instead.')
    trackPageWithTitle = includeTitleInTrackPage
  }

  // ensures Segment write key is present
  if (!prodKey || prodKey.length < 10)
    console.error("segment prodKey must be at least 10 char in length");

  // if dev key is present, ensures it is at least 10 car in length
  if (devKey && devKey.length < 10)
    console.error("if present, devKey must be at least 10 char in length");

  // use prod write key when in prod env, else use dev write key
  // note below, snippet wont render unless writeKey is truthy
  const writeKey = process.env.NODE_ENV === "production" ? prodKey : devKey;

  const loadImmediately = !(delayLoad || delayLoadUntilActivity || manualLoad);
  const reallyTrackPageImmediately = trackPageImmediately && trackPage

  const idempotentPageviewCode = `(function () {
    let lastPageviewPath;
    window.gatsbyPluginSegmentPageviewCaller = function () {
      if (!window.analytics) {
        return
      }${ trackPageOnlyIfReady ? `
      if (!gatsbyPluginSegmentReady) {
        return
      }` : ''}

      let thisPageviewPath = window.location.pathname;
      if (thisPageviewPath === lastPageviewPath) {
        return
      }
      lastPageviewPath = thisPageviewPath;
      window.analytics.page(${ trackPageWithTitle ? 'document.title' : ''});
    };
  })();`;

  let delayedLoadingCode = `(function () {
    let segmentLoaded = false;
    let segmentLoading = false;
    const callbacks = [];
    function safeExecCallback (cb) {
      if (typeof cb === "function") {
        cb();
      }
    }

    window.gatsbyPluginSegmentLoader = function (cb) {
      if (segmentLoaded) {
        safeExecCallback(cb);
        return;
      }

      callbacks.push(cb);
      if (segmentLoading) {
        return;
      }

      segmentLoading = true;

      function loader() {
        if (typeof window.analytics.load !== "function") {
          console.error("Gatsby Plugin Segment: analytics.load is not a function. Has it already been called?");
          return
        }
        gatsbySegmentLoad('${writeKey}');
        segmentLoading = false;
        segmentLoaded = true;
        let cb;
        while ((cb = callbacks.pop()) != null) {
          safeExecCallback(cb);
        }${reallyTrackPageImmediately ? `
        window.gatsbyPluginSegmentPageviewCaller();` : ''}
      };

      if ("requestIdleCallback" in window) {
        requestIdleCallback(loader);
      } else {
        loader();
      }
    };`
    if (delayLoadUntilActivity) {
      delayedLoadingCode += `
    window.addEventListener(
      "scroll",
      function () {
        setTimeout(
          function () {
            window.gatsbyPluginSegmentLoader && window.gatsbyPluginSegmentLoader();
          },
          ${Math.max(0, delayLoadUntilActivityAdditionalDelay || 0)},
        )
      },
      { once: true },
    );`
    }
    if (delayLoad) {
      delayedLoadingCode += `
    setTimeout(
      function () {
        window.gatsbyPluginSegmentLoader && window.gatsbyPluginSegmentLoader();
      },
      ${delayLoadDelay || 1000},
    );`
    }
  delayedLoadingCode += `
  })();`;

  let snippet
  if (customSnippet) {
    snippet = eval('`' + customSnippet + '`')
    snippet += `
    window.analytics.ready(function () {
      gatsbyPluginSegmentReady = true;
    })
    `
  } else {
    // Segment's snippet (version 4.15.3)
    // If this is updated, probably good to update the README to include the version
    snippet = `(function(){var analytics=window.analytics=window.analytics||[];if(!analytics.initialize)if(analytics.invoked)window.console&&console.error&&console.error("Segment snippet included twice.");else{analytics.invoked=!0;analytics.methods=["trackSubmit","trackClick","trackLink","trackForm","pageview","identify","reset","group","track","ready","alias","debug","page","once","off","on","addSourceMiddleware","addIntegrationMiddleware","setAnonymousId","addDestinationMiddleware"];analytics.factory=function(e){return function(){var t=Array.prototype.slice.call(arguments);t.unshift(e);analytics.push(t);return analytics}};for(var e=0;e<analytics.methods.length;e++){var key=analytics.methods[e];analytics[key]=analytics.factory(key)}analytics.load=function(key,e){var t=document.createElement("script");t.type="text/javascript";t.async=!0;t.src="${host}/analytics.js/v1/" + key + "/analytics.min.js";var n=document.getElementsByTagName("script")[0];n.parentNode.insertBefore(t,n);analytics._loadOptions=e};analytics._writeKey="${writeKey}";analytics.SNIPPET_VERSION="4.15.3";`
    if (loadImmediately) {
      snippet += `
    gatsbySegmentLoad('${writeKey}');`
      // Only track if it has been loaded
      if (reallyTrackPageImmediately) {
        snippet += `
    window.gatsbyPluginSegmentPageviewCaller();`
      }
    }
    snippet += `
  }})();`
  }

  const gatsbySegmentLoadStuff = `let gatsbyPluginSegmentReady = ${ trackPageOnlyIfReady ? 'false' : 'true' };
  function gatsbySegmentLoad (writeKey) {
    if (window.analytics && window.analytics.load) {
      window.analytics.load(writeKey);
    }
    gatsbyPluginSegmentReady = true;
  };
  `

  // if `delayLoad` option is true, use the delayed loader
  const pluginCode = `
(function () {
  ${gatsbySegmentLoadStuff}
  ${idempotentPageviewCode}
  ${(delayLoad || delayLoadUntilActivity) && !manualLoad ? delayedLoadingCode : ""}
  ${snippet}
})();`;

  // only render snippet if write key exists
  if (writeKey) {
    setHeadComponents([
      <script
        key="plugin-segment"
        dangerouslySetInnerHTML={{ __html: pluginCode }}
      />,
    ]);
  }
}
